### ISSUE TITLE
Update in Escope 3.2.1 Breaking es6 Spreads

### ISSUE THREAD

**Issue by dtothefp** [eslint/eslint#4481]
Took me a while to find this but started to get an issue with `es6` spreads.

<img width="578" alt="screen shot 2015-11-19 at 4 17 35 pm" src="https://cloud.githubusercontent.com/assets/4656726/11284660/115312ee-8ed9-11e5-9a99-12968f26f071.png">

```js
//login-spec.js
  const {makeGetters, makeActions, Getters, ...store} = loginData;
  const reactor = registerStores(store);
```

```js
//eslint.config.json
{
  "parser": "babel-eslint",
  "plugins": [
    "react"
  ],
  "env": {
    "browser": true
  },
  "ecmaFeatures": {
    "arrowFunctions": true,
    "blockBindings": true,
    "experimentalObjectRestSpread": true,
    "generators": true,
    "jsx": true,
    "module": true,
    "modules": true,
    "restParams": true,
    "spread": true
  }
}
```

Turns out the issues is with `escope` if I install 3.2.0 all works as expected or I guess more accurately as before

<img width="1067" alt="screen shot 2015-11-19 at 4 16 18 pm" src="https://cloud.githubusercontent.com/assets/4656726/11284655/077b4426-8ed9-11e5-934d-3bd2b11c8335.png">

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by michaelficarra** [comment]
0. Those are rests, not spreads, and in particular they are in object binding position.
0. This is an experimental future ECMAScript feature, not supported by eslint.
0. There are many duplicates of this issue, closed, with this same information.

**Comment by nzakas** [comment]
@dtothefp What version of ESLint are you using?

**Comment by dtothefp** [comment]
@nzakas I'm using `1.9.0` but went back to `1.8.0` with no luck. Had to diff my node_modules against a working version in a coworkers local version to find the issue in `escope`. When I constrain `escope` to `3.2.0` in my package.json so eslint uses it rather than it's local version the problem is solved.

Also, put the same issue no `escope` as I'm not sure where this needs to be addressed?

https://github.com/estools/escope/issues/88

**Comment by mysticatea** [comment]
I could not reproduce this with:

```js
{
  // "parser": "babel-eslint", # this is used in the second item.
  "env": {
    "es6": true
  },
  "ecmaFeatures": {
    "experimentalObjectRestSpread": true
  },
  "rules": {
    "no-undef": 2,
    "no-unused-vars": 2,
    "no-redeclare": 2
  }
}
```

```js
const {makeGetters, makeActions, Getters, ...store} = loginData;
const reactor = registerStores(store);
```

## `eslint@1.9.0` and `escope@3.2.1`

```
> eslint test.js

C:\Users\t-nagashima.AD\Documents\GitHub\sand\test.js
  1:8   error  "makeGetters" is defined but never used  no-unused-vars
  1:21  error  "makeActions" is defined but never used  no-unused-vars
  1:34  error  "Getters" is defined but never used      no-unused-vars
  1:55  error  "loginData" is not defined               no-undef
  2:7   error  "reactor" is defined but never used      no-unused-vars
  2:17  error  "registerStores" is not defined          no-undef
```

This is as expected.

## `eslint@1.9.0`, `escope@3.2.1`, and `babel-eslint@4.1.5`

```
> eslint test.js

C:\Users\t-nagashima.AD\Documents\GitHub\sand\test.js
  1:55  error  "loginData" is not defined           no-undef
  2:7   error  "reactor" is defined but never used  no-unused-vars
  2:17  error  "registerStores" is not defined      no-undef
```

This is not as expected, but `no-redeclare` is not appeared.

**Comment by dtothefp** [comment]
Use 'env: { node: true}'

**Comment by mysticatea** [comment]
Hmm, I got the same result with `node: true`.

**Comment by afc163** [comment]
Remove `"parser": "babel-eslint"` is useful for me. :+1:

**Comment by hung-phan** [comment]
Same issue:

1.Eslint 1.9.0
2. Source code and eslint config

Source code
```
import React from 'react';
import SvgIcon from 'material-ui/lib/svg-icon';

function CloseIcon({ iconStyle, style, ...props }) {
  return (
    <SvgIcon viewBox='0 0 48 48' {...props} style={Object.assign({}, style, iconStyle)}>
      <path d='M38 12.83L35.17 10 24 21.17 12.83 10 10 12.83 21.17 24 10 35.17 12.83 38 24 26.83 35.17 38 38 35.17 26.83 24z'/>
    </SvgIcon>
  );
}

export default CloseIcon;
```

Eslint
```
{
  "extends": "airbnb",
  "parser": "babel-eslint",
  "plugins": ["mocha"],
  "env": {
    "mocha": true,
    "browser": true,
    "node": true
  },
  "rules": {
    "comma-dangle": [0, "never"],
    "id-length": [0],
    "jsx-quotes": [2, "prefer-single"],
    "no-console": [1],
    "mocha/no-exclusive-tests": 2
  }
}
```

3. 
![screen shot 2015-11-20 at 2 28 41 pm](https://cloud.githubusercontent.com/assets/3093244/11294081/0b8474f6-8f93-11e5-8b49-5592fd89aa3b.png)

4. I expect will not see this error fired by eslint

**Comment by Uhsac** [comment]
Same issue for me with 1.9.0
This is my configuration file :
```
{
  "extends": "airbnb/base",
  "env": {
    "node": true
  },
  "rules": {
    "func-names": 0,
    "id-length": 0,
    "no-nested-ternary": 0,
    "no-unused-vars": [2, {"args": "none"}],
    "quote-props": [2, "as-needed"],
   },
}
```

And the line that cause the problem : 
`const { errorMessage, errorCode, errorResponse, ...othersProps } = error;`

**Comment by Constellation** [comment]
The fixes are comming in each repository. :)

**Comment by dtothefp** [comment]
@Constellation thanks for investigating. :+1:

**Comment by nzakas** [comment]
Closing, as this was a babel-eslint error.

**Comment by Constellation** [comment]
Published `escope@3.3.0` :)

**Comment by dtothefp** [comment]
@nzakas any plan to bump escope to 3.3.0

**Comment by gyandeeps** [comment]
Reinstall eslint, you would get escope `3.3.0`.

**Comment by dtothefp** [comment]
@Constellation as far as I can tell this issue is still not resolved with `3.3.0`. Made a demo repo here

https://github.com/dtothefp/escope-broken

I am using `babel-eslint` as @nzakas suggests, but as far as I can tell it just references the local `escope` https://github.com/babel/babel-eslint/blob/master/index.js#L40. I'm happy to move this conversation to another repo if it is not appropriate here.

**Comment by smashercosmo** [comment]
Issue is still relevant for 1.10.1

**Comment by nzakas** [comment]
Please follow up with babel-eslint, there's little we can do on our side.

**Comment by raulmatei** [comment]
Hi guys,

It looks like we've had the same issue on some machines that we're running older version of NPM (2.x). Upgrading to 3.3.x seems to solve the issue. I'm not sure how this two relates with each other, We're running latest version of ESLint from Gulp and all other deps are to their latest versions.
